Source/CarthageKit/Project.swift

```diff
@@ -398,20 +398,33 @@ public final class Project {
 	}
 
 	/// Loads the dependencies for the given dependency, at the given version.
-	private func dependencies(for dependency: Dependency<PinnedVersion>) -> SignalProducer<Dependency<VersionSpecifier>, CarthageError> {
+	private func dependencies(for dependency: Dependency<PinnedVersion>, usingCheckout: Bool) -> SignalProducer<Dependency<VersionSpecifier>, CarthageError> {
 
 		switch dependency.project {
 		case .git, .gitHub:
-			let revision = dependency.version.commitish
-			return self.cloneOrFetchDependency(dependency.project, commitish: revision)
-				.flatMap(.concat) { repositoryURL in
-					return contentsOfFileInRepository(repositoryURL, CarthageProjectCartfilePath, revision: revision)
-				}
-				.flatMapError { _ in .empty }
-				.attemptMap(Cartfile.from(string:))
+			let cartfileProducer: SignalProducer<Cartfile, CarthageError>
+			let dependencyCheckoutURL = self.directoryURL.appendingPathComponent(dependency.project.relativePath)
+			 // If the checkout directory doesn't exist, it may be a binary download, so fall back to fetching the cartfile
+			if usingCheckout && FileManager.default.fileExists(atPath: dependencyCheckoutURL.path) {
+				let cartfileURL = dependencyCheckoutURL.appendingPathComponent(CarthageProjectCartfilePath)
+				cartfileProducer = SignalProducer<Cartfile, CarthageError>
+					.attempt {
+						// Ignore errors since the cartfile may not exist if it's a leaf in the dependency graph.
+						return Cartfile.from(file: cartfileURL).recover(with: Cartfile.from(string: ""))
```

Comment:
Won‘t this `recover` also preempt [parsing errors (for example)](https://github.com/Carthage/Carthage/blob/9f1cc9e/Source/CarthageKit/Cartfile.swift#L73)?

Lesser point: just semantically (not sure it has meaningful effect one way or the other), I‘d rather avoid the `Cartfile.from(string: "")` in favor of having the SignalProducer not produce a value (`flatMap`ing `.empty` or the like).